crates/compute/src/cpu/layer.rs

```diff
@@ -124,6 +178,83 @@ impl<T: TowerFamily> ComputeLayer<T::B128> for CpuLayer<T> {
 		}
 		Ok(())
 	}
+
+	fn sum_composition_evals(
+		&self,
+		_exec: &mut Self::KernelExec,
+		log_len: usize,
+		inputs: &[FSlice<'_, T::B128, Self>],
+		composition: &ArithCircuit<T::B128>,
+		batch_coeff: T::B128,
+		accumulator: &mut T::B128,
+	) -> Result<(), Error> {
+		for input in inputs {
+			assert_eq!(input.len(), 1 << log_len);
+		}
+		let ret = (0..1 << log_len)
+			.map(|i| {
+				let row = inputs.iter().map(|input| input[i]).collect::<Vec<_>>();
+				composition.evaluate(&row).expect("Evalutation to succeed")
+			})
+			.sum::<T::B128>();
+		*accumulator += ret * batch_coeff;
+		Ok(())
+	}
+}
+
+// Note: shortcuts for kernel memory so that clippy does not complain about the type complexity in signatures.
+type MemMap<'a, C, F> = KernelMemMap<'a, F, <C as ComputeLayer<F>>::DevMem>;
+type Buffer<'a, C, F> = KernelBuffer<'a, F, <C as ComputeLayer<F>>::DevMem>;
+
+impl<T: TowerFamily> CpuLayer<T> {
+	fn count_total_local_buffer_sizes(
+		mappings: &[MemMap<'_, Self, T::B128>],
+		log_chunk_size: usize,
+	) -> usize {
+		mappings
+			.iter()
+			.map(|mapping| match mapping {
+				KernelMemMap::Chunked { .. } | KernelMemMap::ChunkedMut { .. } => 0,
+				KernelMemMap::Local { .. } => 1 << log_chunk_size,
+			})
+			.sum()
+	}
+
+	fn map_kernel_mem<'a>(
+		mappings: &'a mut [MemMap<'_, Self, T::B128>],
+		local_buffer_alloc: &'a BumpAllocator<T::B128, <Self as ComputeLayer<T::B128>>::DevMem>,
+		log_chunks: usize,
+		i: usize,
+	) -> Vec<Buffer<'a, Self, T::B128>> {
+		mappings
+			.iter_mut()
+			.map(|mapping| match mapping {
+				KernelMemMap::Chunked { data, .. } => {
+					let log_size = checked_log_2(data.len());
+					let log_chunk_size = log_size - log_chunks;
+					KernelBuffer::Ref(<Self as ComputeLayer<T::B128>>::DevMem::slice(
+						data,
+						(i << log_chunk_size)..((i + 1) << log_chunk_size),
+					))
+				}
+				KernelMemMap::ChunkedMut { data, .. } => {
+					let log_size = checked_log_2(data.len());
+					let log_chunk_size = log_size - log_chunks;
```

Comment:
pedantic nit: Should this be `checked_sub()`?

Reply:
Could be. Since this is a private method, I wrote this logic assuming `log_chunks` is chosen in the appropriate range (`KernelMemMap::log_chunks_range`) as a pre-condition.

This call will panic on underflow in debug mode.

Reply:
My reasoning: the caller ensures this is valid and this fn is private to a reference implementation used only in tests, so I guess it's not needed. I can also add it if others think it's needed.